File: src/browser.rb
require 'selenium-webdriver'
class Browser 
  :driver
  CANVAS_CLASS = 'widget-scene-canvas'
  PEGMAN_CLASS = 'widget-expand-button-pegman-icon'

  def initialize() @driver = Selenium::WebDriver.for :chrome end
    
  def move(dist) action.move_to(canvas, vertical_offset, dist).click.perform end
  def drag_pegman() (!pegman || !canvas) ? fin : drag_pegman_to_canvas end
  def go_to(url) @driver.get url end
  def maximize() window.maximize end
  def fin() close; quit end
      
  private def close() @driver.close end
  private def quit() @driver.quit end
  private def action() @driver.action end
  private def drag_pegman_to_canvas() action.drag_and_drop(pegman, canvas).perform end
  private def vertical_offset() window_height.round * 0.05 end
  private def window_height() window.size.height end
  private def window() @driver.manage.window end
  private def canvas() element CANVAS_CLASS end
  private def pegman() element PEGMAN_CLASS end
  private def element(cls) @driver.find_element(class: cls) end
end

File: app/src/main/kotlin/src/Browser.kt
package src

import java.awt.Robot
import org.openqa.selenium.By
import org.openqa.selenium.Keys
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebDriver.Window
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.interactions.Actions

const val CANVAS_ID = "scene"
const val PEGMAN_ID = "q2sIQ"
const val REJECT_COOKIES_XPATH = "//button[normalize-space()='Reject all']"
const val LEFT_ARROW = "//*[@id='compass']/div/button[1]"
const val CENTER_ARROW = "//*[@id='compass']/div/button[2]"
const val RIGHT_ARROW = "//*[@id='compass']/div/button[3]"

enum class Directions {
  LEFT,
  RIGHT,
}

enum class Movements {
  FORWARD,
  BACKWARD,
}

class Browser {
  private val driver: WebDriver = ChromeDriver()
  private val window: Window = driver.manage().window()
  private val actions: Actions = Actions(driver)
  private val robot: Robot = Robot()

  val location: Location?
    get() =
        """@\d+.\d+,-?\d+.\d+,""".toRegex().find(driver.currentUrl)?.let {
          val coords = it.value.split(",")
          return Location(coords[0].split("@")[1].toDouble(), coords[1].toDouble())
        }

  constructor(url: String) {
    Thread.sleep(2000)
    window.fullscreen()
    Thread.sleep(2000)
    driver.navigate().to(url)
    Thread.sleep(5000)
    driver.findElement(By.xpath(REJECT_COOKIES_XPATH)).click()
    Thread.sleep(5000)
    driver.findElement(By.id(PEGMAN_ID)).click()
    driver.findElement(By.id(CANVAS_ID)).click()
    Thread.sleep(3000)
  }

  fun clickWindowCenter(): Unit = actions.click(driver.findElement(By.id(CANVAS_ID))).perform()
  fun forward(): Unit = move(Movements.FORWARD)
  fun backward(): Unit = move(Movements.BACKWARD)
  fun left(): Unit = turn(Directions.LEFT)
  fun right(): Unit = turn(Directions.RIGHT)
  fun finish(): Unit = driver.close().let { driver.quit() }

  private fun move(movement: Movements): Unit =
      when (movement) {
            Movements.FORWARD -> actions.sendKeys(Keys.UP)
            Movements.BACKWARD -> actions.sendKeys(Keys.DOWN)
          }
          .perform()
          .let { Thread.sleep(1000) }

  private fun turn(direction: Directions): Unit =
      actions
          .click(
              when (direction) {
                Directions.LEFT -> driver.findElement(By.xpath(LEFT_ARROW))
                Directions.RIGHT -> driver.findElement(By.xpath(RIGHT_ARROW))
              }
          )
          .perform()
          .let { Thread.sleep(1000) }
}
